src/senaite/patient/browser/widgets/agedob.py

```diff
@@ -19,113 +19,78 @@ class AgeDoBWidget(DateTimeWidget):
-        setattr(context, attr, bool(value))
-
-    def get_current_age(self, dob):
-        """Returns a dict with keys "years", "months", "days"
-        """
-        if not api.is_date(dob):
-            return {}
-
-        delta = patient_api.get_relative_delta(dob)
-        return {
-            "years": delta.years,
-            "months": delta.months,
-            "days": delta.days,
-        }
-
-    def is_age_supported(self, context):
+    def is_age_supported(self):
         """Returns whether the introduction of age is supported or not
         """
         return patient_api.is_age_supported()
 
-    def is_years_only(self, dob):
+    def is_years_only(self):
         """Returns whether months and days are not displayed when the age is
         greater than one year
         """
-        if not patient_api.is_age_in_years():
-            return False
-        dob = self.get_current_age(dob)
-        years = dob.get("years", 0)
-        return years >= 1
+        return patient_api.is_age_in_years()
 
     def process_form(self, instance, field, form, empty_marker=None,
                      emptyReturnsMarker=False, validating=True):
         value = form.get(field.getName())
 
-        # Not interested in the hidden field, but in the age + dob specific
-        if isinstance(value, (list, tuple)):
-            value = value[0] or None
-
         # Allow non-required fields
         if not value:
             return None, {}
 
-        # handle DateTime object when creating partitions
-        if api.is_date(value):
-            self.set_age_selected(instance, False)
-            return value, {}
+        # We always return a dict suitable for the field
+        output = dict.fromkeys(["dob", "from_age", "estimated"], False)
 
-        # Grab the input for DoB first
-        dob = value.get("dob", "")
-        dob = patient_api.to_datetime(dob)
-        age_selected = value.get("selector") == "age"
+        if dtime.is_date(value):
+            # handle date-like objects directly
+            output["dob"] = dtime.to_dt(value)
+            return output, {}
 
-        # remember what was selected
-        self.set_age_selected(instance, age_selected)
+        elif patient_api.is_ymd(value):
+            # handle age-like inputs directly
+            output["dob"] = patient_api.get_birth_date(value)
+            output["from_age"] = True
+            return output, {}
 
-        # Maybe user entered age instead of DoB
-        if age_selected:
-            # Validate the age entered
+        if value.get("selector") == "age":
+            # Age entered
             ymd = map(lambda p: value.get(p), ["years", "months", "days"])
             if not any(ymd):
                 # No values set
-                return None
+                return None, {}
 
             # Age in ymd format
             ymd = filter(lambda p: p[0], zip(ymd, 'ymd'))
             ymd = "".join(map(lambda p: "".join(p), ymd))
```

Comment:
Ok, once again this dance. If it is really required, I would somehow encapsulate it in a `format_ymd` function with a good docstring / comments.

Reply:
Done with e77df12688a1c7590d380afaf7ef165110d47d35 . Is now using to `to_ymd` function